File: backend/app/routers/workspaces.py
"""
Workspaces router — active workspace retrieval and settings updates.
"""

from __future__ import annotations

import logging
import uuid
import re
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.workspaces import Workspace
from app.models.users import User
from app.models.teams import TeamMember
from app.schemas.workspaces import WorkspaceResponse, WorkspaceUpdate

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/workspace", tags=["workspaces"])


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


async def _get_or_create_active_workspace(
    db: AsyncSession,
    user: User,
) -> Workspace:
    """Fetch user's active workspace, fallback to owned/member workspace, or create a default."""
    if user.workspace_id:
        workspace = await db.get(Workspace, user.workspace_id)
        if workspace:
            return workspace

    # Find any owned workspace
    result = await db.execute(
        select(Workspace).where(Workspace.owner_id == user.id).limit(1)
    )
    workspace = result.scalar_one_or_none()
    if workspace:
        user.workspace_id = workspace.id
        db.add(user)
        await db.flush()
        return workspace

    # Find any member workspace
    mem_result = await db.execute(
        select(Workspace)
        .join(TeamMember, TeamMember.workspace_id == Workspace.id)
        .where(TeamMember.user_id == user.id)
        .limit(1)
    )
    workspace = mem_result.scalar_one_or_none()
    if workspace:
        user.workspace_id = workspace.id
        db.add(user)
        await db.flush()
        return workspace

    # Create a default workspace
    workspace_id = uuid.uuid4()
    clean_name = re.sub(r"[^a-zA-Z0-9]", "", user.name.split("@")[0].lower())
    slug = f"workspace-{clean_name}"[:30] or f"workspace-{str(user.id)[:8]}"

    workspace = Workspace(
        id=workspace_id,
        name="My Workspace",
        slug=slug,
        owner_id=user.id,
        plan=user.plan or "free",
        max_members=5 if user.plan == "pro" else 25 if user.plan == "team" else 1,
        settings={
            "defaultModel": "openai/gpt-4o-mini",
            "defaultSystemPrompt": "You are a helpful AI assistant.",
            "description": "My personal AI workspace.",
            "notifications": {
                "emailOnLowBalance": True,
                "emailOnQuotaExceeded": True,
                "emailWeeklyReport": True,
                "emailNewFeatures": False,
                "emailBilling": True,
            }
        },
        created_at=_utcnow(),
        updated_at=_utcnow(),
    )
    db.add(workspace)
    await db.flush()

    member = TeamMember(
        id=uuid.uuid4(),
        workspace_id=workspace.id,
        user_id=user.id,
        role="owner",
        joined_at=_utcnow(),
        created_at=_utcnow(),
    )
    db.add(member)

    user.workspace_id = workspace.id
    db.add(user)
    await db.flush()

    return workspace


@router.get("", response_model=WorkspaceResponse)
async def get_workspace(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Workspace:
    """Fetch the current user's active workspace."""
    return await _get_or_create_active_workspace(db, current_user)


@router.patch("", response_model=WorkspaceResponse)
async def update_workspace(
    payload: WorkspaceUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Workspace:
    """Update active workspace configurations."""
    workspace = await _get_or_create_active_workspace(db, current_user)

    # Validate permission: Owner or Admin only
    if workspace.owner_id != current_user.id:
        tm_res = await db.execute(
            select(TeamMember).where(
                TeamMember.workspace_id == workspace.id,
                TeamMember.user_id == current_user.id,
            )
        )
        member = tm_res.scalar_one_or_none()
        if not member or member.role not in ("owner", "admin"):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only workspace owners or administrators can edit settings.",
            )

    if payload.name is not None:
        workspace.name = payload.name.strip()
        from app.schemas.workspaces import _make_slug
        workspace.slug = _make_slug(workspace.name)

    if payload.settings is not None:
        existing = workspace.settings or {}
        existing.update(payload.settings)
        workspace.settings = existing

    workspace.updated_at = _utcnow()
    db.add(workspace)
    return workspace

File: backend/src/routes/workspaces.ts
import { Router, Request, Response } from "express";
import { requireUser } from "../middleware/auth";
import { prisma } from "../prisma";

const router = Router();

// Retrieve user's current active workspace profile
router.get("/", requireUser, async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ detail: "Unauthorized" });

  let workspaceId = req.user.workspaceId;
  let workspace = null;

  if (workspaceId) {
    workspace = await prisma.workspace.findUnique({
      where: { id: workspaceId },
    });
  }

  // Fallback: look up workspace where user is owner
  if (!workspace) {
    workspace = await prisma.workspace.findFirst({
      where: { ownerId: req.user.id },
    });
  }

  if (!workspace) {
    // Provision workspace in case it is missing
    const defaultSlug = `workspace-${req.user.name.toLowerCase().replace(/[^a-z0-9]/g, "")}-${Date.now().toString().slice(-4)}`;
    workspace = await prisma.$transaction(async (tx) => {
      const newWs = await tx.workspace.create({
        data: {
          name: "My Workspace",
          slug: defaultSlug,
          ownerId: req.user!.id,
          plan: "free",
          maxMembers: 1,
          settings: {
            defaultModel: "openai/gpt-4o-mini",
            defaultSystemPrompt: "You are a helpful AI assistant.",
            description: "My personal AI workspace.",
            notifications: {
              emailOnLowBalance: true,
              emailOnQuotaExceeded: true,
              emailWeeklyReport: true,
              emailNewFeatures: false,
              emailBilling: true,
            },
          },
        },
      });

      // Update user active workspace link
      await tx.user.update({
        where: { id: req.user!.id },
        data: { workspaceId: newWs.id },
      });

      // Add team member owner link
      await tx.teamMember.create({
        data: {
          workspaceId: newWs.id,
          userId: req.user!.id,
          role: "owner",
          joinedAt: new Date(),
        },
      });

      return newWs;
    });
  }

  return res.json(workspace);
});

// Update workspace profile metadata (name, slug, configuration)
router.patch("/", requireUser, async (req: Request, res: Response) => {
  if (!req.user) return res.status(401).json({ detail: "Unauthorized" });

  const workspaceId = req.user.workspaceId;
  if (!workspaceId) {
    return res.status(404).json({ detail: "Active workspace not found" });
  }

  // Ensure user is owner or admin in this workspace
  const member = await prisma.teamMember.findFirst({
    where: { workspaceId, userId: req.user.id },
  });

  if (!member || !["owner", "admin"].includes(member.role)) {
    return res.status(403).json({ detail: "Permission denied. Only owners/admins can modify workspace configurations." });
  }

  const { name, slug, settings, description } = req.body;

  try {
    // If slug is changing, verify uniqueness
    if (slug) {
      const existing = await prisma.workspace.findUnique({ where: { slug } });
      if (existing && existing.id !== workspaceId) {
        return res.status(400).json({ detail: "Workspace URL slug is already taken" });
      }
    }

    const currentWorkspace = await prisma.workspace.findUnique({
      where: { id: workspaceId },
    });

    if (!currentWorkspace) {
      return res.status(404).json({ detail: "Workspace not found" });
    }

    // Merge settings if passed
    let mergedSettings = currentWorkspace.settings || {};
    if (settings) {
      mergedSettings = {
        ...(typeof mergedSettings === "object" ? mergedSettings : {}),
        ...settings,
      };
    }

    if (description) {
      (mergedSettings as any).description = description;
    }

    const updatedWorkspace = await prisma.workspace.update({
      where: { id: workspaceId },
      data: {
        name: name !== undefined ? name : currentWorkspace.name,
        slug: slug !== undefined ? slug : currentWorkspace.slug,
        settings: mergedSettings,
        updatedAt: new Date(),
      },
    });

    return res.json(updatedWorkspace);
  } catch (error: any) {
    console.error("Workspace update failed:", error);
    return res.status(500).json({ detail: error.message });
  }
});

export default router;
